src/Lean/Meta/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,397 @@
+  | /-- the original declaration -/ none
+  | /-- the forward direction of an `iff` -/ mp
+  | /-- the backward direction of an `iff` -/ mpr
+deriving DecidableEq, Inhabited, Ord
+
+instance : ToString DeclMod where
+  toString m := match m with | .none => "" | .mp => "mp" | .mpr => "mpr"
+
+/--
+LibrarySearch has an extension mechanism for replacing the function used
+to find candidate lemmas.
+-/
+@[reducible]
+def CandidateFinder := Expr → MetaM (Array (Name × DeclMod))
+
+namespace DiscrTreeFinder
+
+/-- Add a path to a discrimination tree.-/
+private def addPath [BEq α] (config : WhnfCoreConfig) (tree : DiscrTree α) (tp : Expr) (v : α) :
+    MetaM (DiscrTree α) := do
+  let k ← DiscrTree.mkPath tp config
+  pure <| tree.insertCore k v
+
+/-- Adds a constant with given name to tree. -/
+private def updateTree (config : WhnfCoreConfig) (tree : DiscrTree (Name × DeclMod))
+    (name : Name) (constInfo : ConstantInfo) : MetaM (DiscrTree (Name × DeclMod)) := do
+  if constInfo.isUnsafe then return tree
+  if !allowCompletion (←getEnv) name then return tree
+  withReducible do
+    let (_, _, type) ← forallMetaTelescope constInfo.type
+    let tree ← addPath config tree type (name, DeclMod.none)
+    match type.getAppFnArgs with
+    | (``Iff, #[lhs, rhs]) => do
+      let tree ← addPath config tree rhs (name, DeclMod.mp)
+      let tree ← addPath config tree lhs (name, DeclMod.mpr)
+      return tree
+    | _ =>
+      return tree
+
+/--
+Constructs an discriminator tree from the current environment.
+-/
+def buildImportCache (config : WhnfCoreConfig) : MetaM (DiscrTree (Name × DeclMod)) := do
+  let profilingName := "apply?: init cache"
+  -- Sort so lemmas with longest names come first.
+  let post (A : Array (Name × DeclMod)) :=
+        A.map (fun (n, m) => (n.toString.length, n, m)) |>.qsort (fun p q => p.1 > q.1) |>.map (·.2)
+  profileitM Exception profilingName (← getOptions) do
+    (·.mapArrays post) <$> (← getEnv).constants.map₁.foldM (init := {}) (updateTree config)
+
+/--
+Return matches from local constants.
+
+N.B. The efficiency of this could likely be considerably improved by caching in environment
+extension.
+-/
+def localMatches (config : WhnfCoreConfig) (ty : Expr) : MetaM (Array (Name × DeclMod)) := do
+  let locals ← (← getEnv).constants.map₂.foldlM (init := {}) (DiscrTreeFinder.updateTree config)
+  pure <| (← locals.getMatch  ty config).reverse
+
+/--
+Candidate finding function that uses strict discrimination tree for resolution.
+-/
+def mkImportFinder (config : WhnfCoreConfig) (importTree : DiscrTree (Name × DeclMod))
+    (ty : Expr) : MetaM (Array (Name × DeclMod)) := do
+  pure <| (← importTree.getMatch ty config).reverse
+
+end DiscrTreeFinder
+
+namespace IncDiscrTreeFinder
+
+open LazyDiscrTree (InitEntry createImportedEnvironment)
+
+/--
+The maximum number of constants an individual task performed.
+
+The value below was picked because it roughly correponded to 50ms of work on the machine this was
```

Comment:
```suggestion
The value was picked because it roughly correponded to 50ms of work on the machine this was
```
Because docstrings are mostly consumed outside the context of the source code